### PR TITLE
Add native bag of words support for MLP

### PR DESCRIPTION
**Pull request by achyudh**
The usage of TorchText for learning from bag of words representations was a bottleneck in Hedwig. TorchText is optimized for dealing with plain text and word embeddings, not tf-idf values. It doesn't support compressed input files (and hence we have to store tf-idf values as plain text, leading to files larger than 100 GB), and doesn't seem to be efficient with memory for this case.

This pull request introduces a new trainer, evaluator and data loader classes specifically for dealing with tf-idf representations as sparse matrices. This would allow us to have much larger vocabulary sizes across larger datasets such as IMDB and Yelp.

### CODE REVIEW

File: common/evaluators/bow_evaluator.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                predicted_labels.extend(F.sigmoid(logits).round().long().cpu().detach().numpy())
+                target_labels.extend(labels.cpu().detach().numpy())
+                loss = F.binary_cross_entropy_with_logits(logits, labels.float(), size_average=False)
+            else:
+                predicted_labels.extend(torch.argmax(logits, dim=1).cpu().detach().numpy())
+                target_labels.extend(torch.argmax(labels, dim=1).cpu().detach().numpy())
+                loss = F.cross_entropy(logits, torch.argmax(labels, dim=1))
+
+            if self.args.n_gpu > 1:
+                loss = loss.mean()
```
Comment (daemon) [inline common/evaluators/bow_evaluator.py:61]: I don't think each GPU is guaranteed to have the same batch size in `DataParallel`.

File: models/mlp/model.py [hunk 2]
```diff
@@ -6,10 +6,8 @@ class MLP(nn.Module):
 
     def __init__(self, config):
         super().__init__()
-        dataset = config.dataset
-        target_class = config.target_class
         self.dropout = nn.Dropout(config.dropout)
-        self.fc1 = nn.Linear(dataset.VOCAB_SIZE, target_class)
+        self.fc1 = nn.Linear(config.vocab_size, config.num_labels)
```
Comment (daemon) [inline models/mlp/model.py:10]: Should probably rename MLP to LR in the future.

**Review by daemon — APPROVED** [review]
Minor comments. Approving to not block the pipeline.

File: common/evaluators/bow_evaluator.py (hunk 1 above)
Comment (achyudh) [reply common/evaluators/bow_evaluator.py:61]: Yes, but we do a similar thing in our BERT trainer assuming each GPU has approx the same batch size and it has worked out so far.

File: models/mlp/model.py (hunk 2 above)
Comment (achyudh) [reply models/mlp/model.py:10]: I think we should rename it to perceptron. Since we use sigmoid activation for multilabel, I feel that calling it LR would be misleading.
